Adding Termux Hosting To Docs (#490)
Aiko-IT-Systems/DisCatSharp: DisCatSharp.Docs/articles/misc/hosting.md

Depending on how complex your bot is, you may even consider purchasing a Raspberry Pi ($35).

> # Termux

Reviewer: The heading doesn't really fit

Author: > The headings doesn't really fit

Ah, well no clue.

Reviewer: im going to patch that in a bit, thank you still <3

If you don't have a PC or other gear sitting around, you may use your phone instead. Using [Termux](https://termux.dev/en/) and a program called [proot-distro](https://github.com/termux/proot-distro), we create a Debian virtual machine and configure DotNET to run the bot. For anyone interested, the instructions are detailed below:

## Requirements

> - An Android 7+ phone (5+ is possible but, not recommended as it posses security issues).

Reviewer 2: ```suggestion
- A phone with Android 7 or higher (5+ is possible but, not recommended as it posses security issues).
```

- Termux.
- An internet connection.

## Setup

- Initialize Termux.
```sh
pkg update && pkg upgrade -y
```
>[!NOTE]
> It might ask you for input, just click enter and let the default option execute.

- Install proot-distro package.
```sh
pkg install proot-distro -y
```

- Install a Debian Virtual Machine (VM).
```sh
proot-distro install debian
```
>[!NOTE]
> Installation time for anything will depend on your internet speed.

- Login into Debian and initialize Debian.
```sh
proot-distro login debian
```
```sh
apt update -y && apt upgrade -y
```

- Install git and wget.
```sh
apt install git wget -y
```

- We get the DotNET script.
```sh
wget https://dot.net/v1/dotnet-install.sh -O dotnet-install.sh && chmod +x ./dotnet-install.sh
```

- Now, we install the DotNET sdk and runtime.
```sh
./dotnet-install.sh --channel 8.0
```
>[!NOTE]

> > The script doesn't automatically add DotNET to path so run, `export PATH=$PATH:$HOME/.dotnet` and keep it in your `.bashrc` file which can be found in your `~` directory. To edit the `.bashrc` file run, `nano ~/.bashrc` and paste in the arguments. The `.bashrc` file is your configuration file for the bash shell environment.

Reviewer 2: ```suggestion
> If the script doesn't automatically add `dotnet` to the path environment variable, run `export PATH=$PATH:$HOME/.dotnet` and add this command to the end of your `.bashrc` file which can be found in your user directory. To edit the `.bashrc` file run, `nano ~/.bashrc` and paste in the arguments. The `.bashrc` file is your configuration file for the bash shell environment.
```

>[!WARNING]

> > You will get an error code `0x8007000E` when you try to run it to fix it run, `export DOTNET_GCHeapHardLimit=1C0000000` and also put that in your `~/.bashrc` file. You may also get a huge error message about a "valid ICU package" to fix that run, `export DOTNET_SYSTEM_GLOBALIZATION_INVARIANT=1` and run the bot again. Or simply add these lines to your `.csproj` file in your project folder:

Reviewer 2: ```suggestion
> You might encounter the error code `0x8007000E` when you try to run the bot. To fix it, run `export DOTNET_GCHeapHardLimit=1C0000000` and also put that command into your `~/.bashrc` file. 

You'll possibly also get hit with an error message about a "valid ICU package" to fix that, run `export DOTNET_SYSTEM_GLOBALIZATION_INVARIANT=1` and run the bot again. Or simply add these lines to your `.csproj` file in your project folder:
```

```xml
<RuntimeHostConfigurationOption Include="System.Globalization.Invariant" Value="true" />
```

### Cloud Hosted Code

- If you have your code hosted with git or simply are able to download it, get the source code and `cd` into the directory. Example of a git repository on GitHub:
```sh
git clone https://github.com/username/repo.git
cd repo
```
- Add your configuration file then build the project.

### Local Source Code

> - If you do not have your code hosted with git let's move the code inside the VM by following the given steps: 

Reviewer 2: ```suggestion
- If you do not have your source code hosted remotely, you can move the source code inside the VM by following the given steps: 
```

- Enable developer options on your phone, find out [here](https://developer.android.com/studio/debug/dev-options) and also enable USB debugging which you can do by going inside `Settings > Developer options > USB debugging` or where ever your phone's is located.

- Find the option `Default USB configuration` and choose `Transferring files`. Disconnect your phone from the usb and reconnect it again.

> [!NOTE]

> > A prompt should come up saying a new device is connected to the PC which is the phone. Go to `This PC` on Windows and double click on your device.

Reviewer 2: I don't think this note is necessary

Author: Alright

- Double click on `Phone` or what it is named, then simply copy your project folder from your pc and paste it in there. If you want to keep it in a custom path, remember the path to the project.

>[!NOTE]

> > Transferring of files will take a bit and be sure to not interrupt the process by shutting down any devices or disconnecting.

Reviewer 2: ```suggestion
> Transferring of files might take a bit. Be sure to not interrupt the process by shutting down or disconnecting your phone
```

- First, we setup Termux storage run, `termux-setup-storage` and then it'll prompt you to give Termux access to your files and folders. Then run, `ls ~/storage/shared` to make sure you have access.

>[!WARNING]
> If you do not give Termux access to your files and folders, nothing will work going from here on.

- Now, exit out of Debian and into normal Termux and make a new environment variable which leads us to the VM's root directory. Run, `export PROOTDISTROFS=/data/data/com.termux/files/usr/var/lib/proot-distro/installed-rootfs` and run `source $HOME/.bashrc`.

>[!NOTE]

> >  If you do not have a `.bashrc` in your `$HOME` just run, `nano $HOME/.bashrc`, save and exit. Then run, the source command again.

Reviewer 2: ```suggestion
> If you don't have a `.bashrc` file in your $HOME, just run `nano $HOME/.bashrc`, save, and exit. Then run the source command again.
```

> - Now, let's ove over the source code from our phone to the VM. Simply, run, `mv projectname/ $PROOTDISTROFS/debian/root/`.

Reviewer 2: Typo and let's sounds a bit weird here
```suggestion
- Now, we'll move over the source code from our phone to the VM. Simply, run, `mv projectname/ $PROOTDISTROFS/debian/root/`.
```

Author: Yeah, let's just go with your changes and call it :D

>[!NOTE]

> > Double, check for file path's as you may accidentally move the wrong files around.

Reviewer 2: Grammar

```suggestion
> Double-check file paths, as you might accidentally move the wrong files around.
```

> - Now, let's login back to Debian `proot-distro login debian` and check if it's here. To check, run `ls` and if you need `projectname` in the list then it's there. From here.

Reviewer 2: ```suggestion
- Now, let's log back into Debian `proot-distro login debian` and check if the files are here. To check, you can run `ls` to list the current content of the folder.
```

- Add your configuration file then build the project.

>[!NOTE]

> > If you have folders such as `bin/` and `obj/` from your prior builds on Windows, delete those by running, `rm -rf bin/ obj/` as without doing so, will result in an error. 

Reviewer 2: ```suggestion
> If you have folders such as `bin/` and `obj/` from your prior builds on your PC, delete those by running `rm -rf bin/ obj/`. You might run into issues otherwise.
```

## Profit

> Bot should be working fine, given you follow appropriate steps. For support or any inquires you can join the DisCatSharp [Discord](https://discord.com/invite/2HWta4GXus).

Reviewer: I don't know if we want to support this method officially.
Termux is a cool suggestion, but in any case a real "vps" would be better.

Author: > I don't know if we want to support this method officially. Termux is a cool suggestion, but in any case a real "vps" would be better.

Well, I just wanted to add a cheaper alternative as to a VPS or something since, if you have an old phone lying around it's just way cheaper and convenient for small or local scaled bots.

Depending on how complex your bot is, you may even consider purchasing a Raspberry Pi ($35).

# Termux

> For those who don't have some PC or other hardware lying around alternatively you could use your phone. Using [Termux](https://termux.dev/en/) and a package called [proot-distro](https://github.com/termux/proot-distro), which we use to install a Debian vm and setup DotNET to run the bot. To those interested the steps are given below:

Reviewer 2: > don't have some PC -> don't have a PC

Second sentence is kind of a mess, please rephrase.

Author: Alright

## Requirements

- An Android 5+
- Termux
- An internet connection

## Setup

- Initialize Termux
```sh
pkg update && pkg upgrade -y
```
*Note, it might ask you for input, just click enter and let the default option execute*

- Install proot-distro package
```sh
pkg install proot-distro -y
```

> - Install a Ubuntu vm

Reviewer 2: I feel vm should be typed out fully, for clarity.
Also as discussed on Discord, please look into using Debian instead because it's easier to run.

Author: Got it, looking into Debian currently

```sh

> proot-distro install ubuntu -y

Reviewer 2: `-y` is invalid.

Author: Ah, yeah hah

```
*Note, it may take a bit depending on your internet speed*

- Login into Ubuntu and initialize Ubuntu
```sh
proot-distro login ubuntu
```
```sh

> apt-get update -y && apt-get upgrade -y

Reviewer 2: From my knowledge, apt-get is supposed to be used for scripts and such. At the moment it doesn't really make a difference but to future proof: Use apt instead, especially because apt 3.0 actually makes the output a lot more usable when that's released.

Author: Alright, noted also cannot find a way to.update to apt 3.0, if you know how please provide information

```
*Note, it may take a bit depending on your internet speed*

- Update and install git and sdk
```sh

> apt-get update -y && apt-get install -y dotnet-sdk-8.0 git -y

Reviewer 2: I think, in most cases, actually compiling the bot on a mobile device wouldn't be the best idea so maybe just downloading the runtime is enough. Whether you want to change this is up to you, just a thought.

Author: Alright, will look into that, thanks for feedback

```
*Note, will take a bit*

- Clone the repo where your bot source code is and cd into it
```sh

> git clone https://github.com/KristalliDev/alpine.git

Reviewer 2: I don't think we should add a reference to a repo outside of Aitsys' control.
Please find an alternative to this step, maybe describing how to get the user's own bot onto this vm.

Author: Alright, noted also currently I cannot find a way to update to apt 3.0, if you know how please do provide some information

Author: Alright, I'll just replace my username and repo with "username"/"repo"

cd alpine
```
- Add your configuration file and run the project
```sh
dotnet run
```

> **Note, you may get an error code `0x8007000E` to fix type `export DOTNET_GCHeapHardLimit=1C0000000` and also put that in your `.bashrc`**

Reviewer 2: This note needs to be more clear.

Also notes like these have their own format, see https://github.com/Aiko-IT-Systems/DisCatSharp/blob/main/DisCatSharp.Docs/articles/modules/audio/lavalink_v4/setup.md?plain=1#L24 for how notes like these are done. Also likely more important than note, warn might be more appropriate.

Author: Noted.

## Profit

> Bot should be working fine, given you follow appropriate steps. For support or any inquires you can join the DisCatSharp [Discord](https://discord.com/invite/2HWta4GXus) or create an [Issue](https://github.com/Aiko-IT-Systems/DisCatSharp/issues).

Reviewer 2: Discord should be used for support, issues for bug reports/feature requests.

Author: Noted.

>[!WARNING]
> You might encounter the error code `0x8007000E` when you try to run the bot. To fix it, run `export DOTNET_GCHeapHardLimit=1C0000000` and also put that command into your `~/.bashrc` file. 

> >> You'll possibly also get hit with an error message about a "valid ICU package" to fix that, run `export DOTNET_SYSTEM_GLOBALIZATION_INVARIANT=1` and run the bot again. Or simply add these lines to your `.csproj` file in your project folder:

Reviewer 2: ```suggestion
> You'll possibly also get hit with an error message about a "valid ICU package" to fix that, run `export DOTNET_SYSTEM_GLOBALIZATION_INVARIANT=1` and run the bot again. Or simply add these lines to your `.csproj` file in your project folder:
```
